plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/ForceApply.hs

```diff
@@ -0,0 +1,189 @@
+             (delay^|T1|
+               (\X1 ->
+                  delay^|T2| (\X2 ->
+                    delay^|T3| (\X3 ->
+                      ... ->
+                        delay^|Tn| (\Xn -> body)))))
+            ) X1)) X2)) X3) ...)) Xn
+ @
+ In the end, after applying the "UntypedPlutusCore.Transform.ForceDelay" optimisation:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((force^|T2|
+           ((\X1 ->
+             delay^|T2| (\X2 ->
+               delay^|T3| (\X3 ->
+                 ... ->
+                   delay^|Tn| (\Xn -> body)))))
+        X1) X2)) X3 ...)) Xn
+ @
+ Notice that the next two reduction steps (applying @X1@ and reducing @force (delay ...)@)
+ produce an equivalent term to applying the transformation and then the reduction rule
+ for application.
+ This is easy to check, so we continue by showing what the "optimised term" looks like:
+ @
+   (force^|Tn|
+     (...
+       (force^|T3|
+         ((\X1 -> \X2 ->
+           delay^|T3| (\X3 ->
+             ... ->
+               delay^|Tn| (\Xn -> body))
+        X1) X2)) X3 ...)) Xn
+ @
+ The term can be optimised further by "erasing" the @force^|T3|@ and @delay^|T3|@ pair,
+ and so on until @Tn@.
+
+ For examples of terms we can optimise, see the test cases in the
+ "Transform.Simplify.forceApply*" module of the test suite.
+-}
+{-# LANGUAGE LambdaCase #-}
+module UntypedPlutusCore.Transform.ForceApply
+    ( forceApply
+    ) where
+
+import UntypedPlutusCore.Core
+
+import Control.Lens (transformOf)
+import Control.Monad (guard)
+import Data.Foldable (foldl')
+import Data.Maybe (fromMaybe)
+
+{- | Traverses the term, for each node applying the optimisation
+ detailed above. For implementation details see 'optimisationProcedure'.
+
+ Note: this optimisation should be applied after
+ 'UntypedPlutusCore.Transform.ForceDelay.forceDelay' for reasons mentioned
+ above.
+-}
+forceApply :: Term name uni fun a -> Term name uni fun a
+forceApply = transformOf termSubterms processTerm
+
+{- | Checks whether the term is of the right form, and "pushes"
+ the 'Force' down into the underlying lambda abstractions.
+-}
+processTerm :: Term name uni fun a -> Term name uni fun a
+processTerm = \case
+    original@(Force _ subTerm) ->
+        fromMaybe original (optimisationProcedure subTerm)
+    t -> t
+
+{- | Converts the subterm of a 'Force' into specialised types for representing
+ multiple applications on top of multiple abstractions. Checks whether the lambda
+ will eventually get "exactly reduced" and applies the optimisation.
+ Returns 'Nothing' if the optimisation cannot be applied.
+-}
+optimisationProcedure :: Term name uni fun a -> Maybe (Term name uni fun a)
+optimisationProcedure term = do
+    asMultiApply <- toMultiApply term
+    innerMultiAbs <- toMultiAbs . toApply $ asMultiApply
+    guard $ length (applyToIter asMultiApply) == length (absVars innerMultiAbs)
+    case rhs innerMultiAbs of
+        Delay _ subTerm ->
```

Comment:
> We still insist on encountering a Delay here even though it's not really necessary? Or do you disagree that it's not necessary?

For this branch in the `case` it is necessary because we're removing the `Force`.

You do have a point I didn't consider: what happens if instead of returning `Nothing` in the non-delay case we would push the `Force` down? (Which I think is what you were trying to tell me but I thought it meant "keep the algorithm the same but don't match on `Delay`") .

A quick look at this, unfortunately, makes a few of the benchmark test outputs a bit worse. For example, I am getting the following diff for `checkScriptContextEqualityTerm-20`:
```
-({cpu: 304395100
 -| mem: 1215868})

+({cpu: 304533100
+| mem: 1216468})
```
No changes to any golden term files.

I think, though, that what you suggest here is part of a different class of optimisations. As we discussed during the technical meeting, I think the "pushing T down without eliminating it" where T is either Force, Constr, or something else for which it's sound to "float lets out of" in UPLC, is a different optimisation altogether. I think we need a spike for investigating them. After that I will be much more confident to say if we can somehow merge all of them into a single optimisation.

> This pass doesn't do the job of eliminating a Force-Delay pair if they're next to each other, but eliminates them if they're a few Apply-LamAbs pairs apart. 

I can move the force-delay cancelling into this module, that's fine and after thinking about it a good idea because it's essentially the case where the apply is empty and the lambda is empty. :+1: 

> I'd say pick one: either you always eliminate Force-Delay pairs or you let the other pass clean them up
Could doing the job of the force-delay pass here make not only allow us to remove the latter, but also simplify this pass as well? You have that Maybe indirection in toMultiApply and toMultiAbs -- what if you simply removed the Maybe, would it still result in correct logic despite providing more functionality and requiring less code?

That wouldn't work. You're saying that this pass should only push forces down, correct? Say you have: 
```
force (force [(\x1 -> delay [(\x2 -> delay [(\x3 -> body) 5]) 7]) 9])
==> (push inner force)
force [(\x1 -> force (delay [(\x2 -> delay [(\x3 -> body) 5]) 7]) 9]
==> (push outer force)
[(\x1 -> force (force (delay [(\x2 -> delay [(\x3 -> body) 5]) 7])) 9]
```
...and you get stuck, because you no longer have a `force` applied to an application of a lambda.

Let's say that then you apply a different pass which removes forces immediately followed by delays, you get:
```
[(\x1 -> force [(\x2 -> delay [(\x3 -> body) 5]) 7]) 9]
```
Now you need to run the "push" pass again!

**My point is that being able to push multiple forces in _one_ pass requires that you also remove the force-delay pairs along the way.**

Otherwise you have to do an arbitrary number of traversals of the term, which we agreed is not ideal. This is the reason why we can't generalise this optimisation further into some general "let-floating" algorithm, and why we _need_ to have something separate for `force` and `delay` interleaving.

I hope this makes sense, let me know what you think.

Reply:
> A quick look at this, unfortunately, makes a few of the benchmark test outputs a bit worse. For example, I am getting the following diff for `checkScriptContextEqualityTerm-20`:

That is weird... Thanks a lot for doing the analysis, could you please mention this in the comments in the code, so that we don't forget why we didn't implement the more general algorithm?

> I think, though, that what you suggest here is part of a different class of optimisations. As we discussed during the technical meeting, I think the "pushing T down without eliminating it" where T is either Force, Constr, or something else for which it's sound to "float lets out of" in UPLC, is a different optimisation altogether.

I think it's the same optimization, but YMMV.

> My point is that being able to push multiple forces in one pass requires that you also remove the force-delay pairs along the way.

Great analysis, thank you for elaborating, I now see that what I proposed doesn't make much sense.